ratelimits/gcra.go

```diff
@@ -0,0 +1,119 @@
+package ratelimits
+
+import (
+	"math"
+	"time"
+
+	"github.com/jmhodges/clock"
+)
+
+// divThenRound divides two int64s and rounds the result to the nearest integer.
+// This is used to calculate request intervals and costs in nanoseconds.
+func divThenRound(x, y int64) int64 {
+	return int64(math.Round(float64(x) / float64(y)))
+}
+
+// maybeSpend uses the GCRA algorithm to decide whether to allow a request. It
+// returns a Decision struct with the result of the decision and the updated
+// TAT. The cost must be 0 or greater and <= the burst capacity of the limit.
+func maybeSpend(clk clock.Clock, rl limit, tat time.Time, cost int64) *Decision {
+	if cost < 0 || cost > rl.Burst {
+		// The condition above is the union of the conditions checked in Check
+		// and Spend methods of Limiter. If this panic is reached, it means that
+		// the caller has introduced a bug.
+		panic("invalid cost for maybeSpend")
+	}
+	nowUnix := clk.Now().UnixNano()
+	tatUnix := tat.UnixNano()
+
+	// If the TAT is in the future, use it as the starting point for the
+	// calculation. Otherwise, use the current time. This is to prevent the
+	// bucket from being filled with capacity from the past.
+	if nowUnix > tatUnix {
+		tatUnix = nowUnix
+	}
+
+	// Compute the cost increment.
+	costIncrement := rl.emissionInterval * cost
+
+	// Deduct the cost to find the new TAT and residual capacity.
+	newTAT := tatUnix + costIncrement
+	difference := nowUnix - (newTAT - rl.burstOffset)
+
+	var residual int64
+	if difference < 0 {
+		// Too little capacity to satisfy the cost, deny the request.
+		residual = divThenRound(nowUnix-(tatUnix-rl.burstOffset), rl.emissionInterval)
```

Comment:
No need to pre-declare residual, since it can just be scoped to this if-clause that is about to return.
```suggestion
	// Deny the request if the newTAT would be more than burstOffset in the future.
	if difference < 0 {
		residual := divThenRound(nowUnix-(tatUnix-rl.burstOffset), rl.emissionInterval)
```